hoomd/write/hdf5.py

```diff
@@ -0,0 +1,295 @@
+
+    def __init__(self, filename, logger, mode="a"):
+        if h5py is None:
+            raise ImportError(f"{type(self)} requires the h5py pacakge.")
+        param_dict = ParameterDict(filename=typeconverter.OnlyTypes(
+            (str, PurePath)),
+                                   logger=logging.Logger,
+                                   mode=str)
+        if (rejects := self._reject_categories
+                & logger.categories) != logging.LoggerCategories["NONE"]:
+            reject_str = logging.LoggerCategories._get_string_list(rejects)
+            raise ValueError(f"Cannot have {reject_str} in logger categories.")
+        param_dict.update({
+            "filename": filename,
+            "logger": logger,
+            "mode": mode
+        })
+        self._param_dict = param_dict
+        self._fh = None
+        self._attached_ = False
+
+    def _initialize(self, communicator):
+        if communicator is None or communicator.rank == 0:
+            self._fh = h5py.File(self.filename, mode=self.mode)
+        else:
+            self._fh = None
+        self._validate_scheme()
+        self._frame = self._find_frame()
+
+    def __del__(self):
+        """Closes file upon destruction."""
+        if getattr(self, "_fh", None) is not None:
+            self._fh.close()
+
+    def _setattr_param(self, attr, value):
+        """Makes self._param_dict attributes read only."""
+        raise ValueError("Attribute {} is read-only.".format(attr))
+
+    def attach(self, simulation):
+        self._initialize(simulation.device.communicator)
+        self._attached_ = True
+
+    def _attached(self):
+        return self._attached_
+
+    def detach(self):
+        self._attached_ = False
+        if self._fh is not None:
+            self._fh.close()
+
+    def act(self, timestep):
+        """Write a new frame of logger data to the HDF5 file."""
+        log_dict = util._dict_flatten(self.logger.log())
+        if self._fh is None:
+            return
+        if self._frame == 0:
+            self._initialize_datasets(log_dict)
+        for key, (value, category) in log_dict.items():
+            if logging.LoggerCategories[category] in self._reject_categories:
+                continue
+            if value is None:
+                continue
+            str_key = "/".join(("hoomd-data",) + key)
+            if str_key not in self._fh:
+                raise RuntimeError(
+                    "The logged quantities cannot change within a file.")
+            dataset = self._fh[str_key]
+            dataset.resize(self._frame + 1, axis=0)
+            dataset[self._frame, ...] = value
+        self._frame += 1
+        self._fh["hoomd-data"].attrs["frames"] = self._frame
+
+    @_skip_fh
+    def flush(self):
+        """Write out all data currently buffered in memory.
+
+        Without calling this, data may be stored in the h5py.File object without
+        being written to disk yet.
+        """
```

Comment:
```suggestion
        .. rubric:: Examples
        
        Flush one writer:
        
        .. skip: if(h5py_not_available)
        
        .. code-block:: python
        
            hdf5_log_writer.flush()
            
        Flush all write buffers:
        
        .. skip: if(h5py_not_available)
        
        .. code-block:: python
        
            for writer in simulation.operations.writers:
                if hasattr(writer, 'flush'):
                    writer.flush()
        """
```
Will need to move the invisible setup code up higher in the document for this to pass tests.